Support %-based string formatting

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45976 Support %-based string formatting**

Differential Revision: [D24374215](https://our.internmc.facebook.com/intern/diff/D24374215)

Commit:
Support %-based string formatting

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3136,6 +3136,14 @@ struct to_ir {
+        if (tree->tree(0)->kind() == TK_STRINGLITERAL) {
+          auto values = getValues(tree->trees(), /*maybe_unpack=*/false);
+          auto node = graph->create(aten::percentFormat, values, 1)
+                        ->setSourceRange(tree->range());
```
I think you need to set node output type here.

Commit:
Update on "Support %-based string formatting"




[ghstack-poisoned]

Commit:
Update on "Support %-based string formatting"




[ghstack-poisoned]

Commit:
Update on "Support %-based string formatting"




[ghstack-poisoned]

gmagogsfm reviewed (changes requested):
Reviewed part of it, please address current comments and we can iterate.

ansley commented on c10/util/C++17.h:
```diff
@@ -6,6 +6,7 @@
 #include <utility>
 #include <memory>
 #include <sstream>
+#include <iostream>
```
Not an accident, actually. I need to include iostream to format the sstream in `torch/csrc/jit/runtime/vararg_functions.cpp`

I assumed that you put the include directives together in certain files instead of just tossing them in whatever file you were working on. I tried to trace that "certain file", and it looked like this was where other important includes were, so this is where I put mine. Where should I put the `#include <iostream>` instead?

ansley replied:
Thank you!

ansley commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3136,6 +3136,16 @@ struct to_ir {
             tree->range(), *method.graph(), kind, named_values, {});
       }
+      case '%': {
+        if (tree->tree(0)->kind() == TK_STRINGLITERAL) {
```
Could you give me an example test case? I'm looking at possible tokens in `lexer.h`, and I don't see what else it could be. (`TK_STRING` is reserved for atoms.)

gmagogsfm replied:
I was thinking about something like:

def fn(x: str):
    return x % "hi"

Commit:
Update on "Support %-based string formatting"




[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/runtime/vararg_functions.cpp:
```diff
@@ -39,6 +39,108 @@ void format(Stack& stack, size_t num_inputs) {
+        ival.isScalar(),
+        "%",
+        key,
+        " format: A number is required, not ",
```
Conventionally, error message should be something like
XXX is required *got* YYY

ansley replied:
Thanks, I'll change it! I was copying Python's error message.

gmagogsfm commented on torch/csrc/jit/runtime/vararg_functions.cpp:
```diff
@@ -39,6 +39,108 @@ void format(Stack& stack, size_t num_inputs) {
+std::string getFormattedArg(char key, const IValue& ival, int precision=6) {
+  //TODO: Implement precison-based formatting
+  std::stringstream ss;
+  bool added = false;
```
Using `added` is a bit confusing, I think having `ss << ival` in every case is perfectly fine.

gmagogsfm commented on torch/csrc/jit/runtime/vararg_functions.cpp:
```diff
@@ -39,6 +39,108 @@ void format(Stack& stack, size_t num_inputs) {
+}
+
+void percentFormat(Stack& stack, size_t num_inputs) {
+  std::vector<char> specifiers = {'d', 'i', 'e', 'E', 'f', 'F', 's', 'c'};
```
I think this can be made `const static`

gmagogsfm replied:
and make it into an `unordered_set` would make your check on line 132 much faster

gmagogsfm replied:
In fact, thinking more about it, this check isn't exactly necessary. The helper method needs to check validity of format char anyway, it can return whether the format is valid or not.

gmagogsfm commented on torch/csrc/jit/runtime/vararg_functions.cpp:
```diff
@@ -39,6 +39,108 @@ void format(Stack& stack, size_t num_inputs) {
+  if (!added) {
+    ss << ival;
+  }
+  return ss.str();
```
This seems a bit wasteful, what you can do is pass `stringstream` object reference to this helper function and directly stream the results to it.

gmagogsfm commented on torch/csrc/jit/runtime/vararg_functions.cpp:
```diff
@@ -39,6 +39,108 @@ void format(Stack& stack, size_t num_inputs) {
+  std::vector<char> specifiers = {'d', 'i', 'e', 'E', 'f', 'F', 's', 'c'};
+  auto format = peek(stack, 0, num_inputs).toStringRef();
+  auto args = last(stack, num_inputs - 1)[0];
+  auto args_size = !args.isTuple() ? 1 : args.toTuple()->elements().size();
```
Add a TORCH_CHECK to ensure that there is only 1 input when its type is tuple.

gmagogsfm commented on torch/csrc/jit/runtime/vararg_functions.cpp:
```diff
@@ -39,6 +39,108 @@ void format(Stack& stack, size_t num_inputs) {
+  size_t used_args = 0;
+  for (size_t begin = 0; true; ++used_args) {
+    size_t loc = format.find('%', begin);
+    if (loc >= format.length() - 1) {
```
I think a more idiomatic way to check if a token is found is to check against string::npos

http://www.cplusplus.com/reference/string/string/npos/

gmagogsfm commented on torch/csrc/jit/runtime/vararg_functions.cpp:
```diff
@@ -39,6 +39,108 @@ void format(Stack& stack, size_t num_inputs) {
+    char key = format.at(loc + 1);
+    TORCH_CHECK(std::find(specifiers.begin(), specifiers.end(), key) !=
+        specifiers.end(), "The specifier ", key, " is not supported in TorchScript");
+    auto arg = !args.isTuple() ? args : args.toTuple()->elements()[used_args];
```
I think it is worth making `args.isTuple() ? args : args.toTuple()->elements()` a temp variable to make the code cleaner.

gmagogsfm commented on torch/csrc/jit/runtime/vararg_functions.cpp:
```diff
@@ -39,6 +39,108 @@ void format(Stack& stack, size_t num_inputs) {
+  auto args_size = !args.isTuple() ? 1 : args.toTuple()->elements().size();
+  std::stringstream ss;
+  size_t used_args = 0;
+  for (size_t begin = 0; true; ++used_args) {
```
This should really be turned into a `while` loop.

gmagogsfm commented on test/test_jit.py:
```diff
@@ -2581,6 +2581,120 @@ def forward(self, x):
 
 
 class TestScript(JitTestCase):
+
```
With so many test case added, I think it is worth to make its own test file, search for `test_enum.py` for an example.

Commit:
Update on "Support %-based string formatting"




[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3136,6 +3136,19 @@ struct to_ir {
+          output->setType(StringType::get());
+          return output;
+        }
+      }
```
I sense that the fallthrough behavior here is intentional, but still better to make it obvious by adding a comment here.

In c++ 17, there is a fallthrough attribute where you can make it really obvious, but Pytorch hasn't adopted C++ 17 yet.

gmagogsfm commented on torch/csrc/jit/runtime/vararg_functions.cpp:
```diff
@@ -39,6 +39,116 @@ void format(Stack& stack, size_t num_inputs) {
+      TORCH_CHECK(ival.isInt() || (ival.isString() && ival.toStringRef().length() == 1), "Got ", ival.tagKind(), ", but an int or char is required for formatting");
+      if (ival.isInt()) {
+        ss << static_cast<char>(ival.toInt());
+      }
```
Nit:

unsure why formatter is doing this, but `}` and `else` should be on same line generally

Commit:
Update on "Support %-based string formatting"




[ghstack-poisoned]

Commit:
Update on "Support %-based string formatting"




[ghstack-poisoned]

gmagogsfm commented on test/jit/test_string_formatting.py:
```diff
@@ -0,0 +1,149 @@
+        with self.assertRaisesRegex(RuntimeError, "Incomplete format specifier"):
+            @torch.jit.script
+            def fn(arg1: str) -> str:
+                return "%s in template %" % arg1
```
I think you can use 
```
# noqa: F501
```

to suppress flake8 warning

gmagogsfm commented on test/jit/test_string_formatting.py:
```diff
@@ -0,0 +1,149 @@
+        with self.assertRaisesRegex(RuntimeError, "Too many arguments for format string"):
+            @torch.jit.script
+            def fn(arg1: str, arg2: str) -> str:
+                return "%s in template" % (arg1, arg2)
```
Similar here, use
```
# noqa: F507
```

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3136,6 +3136,20 @@ struct to_ir {
+      case '%': {
+        auto lhs = emitSugaredExpr(Expr(tree->tree(0)), 0)
+                       ->asValue(tree->tree(0)->range(), method);
+        auto lhs_type = lhs->type();
```
Please fix clang-tidy error here

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3136,6 +3136,20 @@ struct to_ir {
+          output->setType(StringType::get());
+          return output;
+        }
+        // else, fall through to the next case
```
Thinking more about it, I am really nervous about this fallthrough case since it can be easily broken. Could you refactor code that handles the next case into a function, call that function here and don't fallthrough?

gmagogsfm commented on torch/csrc/jit/runtime/register_special_ops.cpp:
```diff
@@ -431,9 +431,7 @@ RegisterOperators reg({
         aliasAnalysisConservative()),
     Operator(
         "aten::set_grad_enabled(bool val) -> ()",
-        [](Stack* stack) {
```
auto-formatted?

gmagogsfm commented on torch/csrc/jit/runtime/vararg_functions.cpp:
```diff
@@ -39,6 +39,130 @@ void format(Stack& stack, size_t num_inputs) {
+    char key,
+    const IValue& ival,
+    std::stringstream& ss,
+    int precision = 6) {
```
Please fix clang-tidy issue.

gmagogsfm commented on torch/csrc/jit/runtime/vararg_functions.cpp:
```diff
@@ -39,6 +39,130 @@ void format(Stack& stack, size_t num_inputs) {
+  // TODO: Implement precison-based formatting
+  switch (key) {
+    case 'd':
+    case 'i': {
```
I think these curly brackets are not necessary, they are only needed when you create new variables inside `case`.

gmagogsfm commented on torch/csrc/jit/runtime/vararg_functions.cpp:
```diff
@@ -39,6 +39,130 @@ void format(Stack& stack, size_t num_inputs) {
+          ", but a number is required for formatting");
+      ss << std::setprecision(precision) << std::scientific;
+      if (key == 'E') {
+        ss << std::uppercase;
```
We should probably reset uppercase printing when getting out of this case.

gmagogsfm commented on torch/csrc/jit/runtime/vararg_functions.cpp:
```diff
@@ -39,6 +39,130 @@ void format(Stack& stack, size_t num_inputs) {
+          "Got ",
+          ival.tagKind(),
+          ", but a number is required for formatting");
+      ss << std::setprecision(precision) << std::fixed;
```
I don't think it matters *now*, but in general we should reset state like "precision" of printing stream when we are done in this case.

gmagogsfm commented on torch/csrc/jit/runtime/vararg_functions.cpp:
```diff
@@ -39,6 +39,130 @@ void format(Stack& stack, size_t num_inputs) {
+    ss << format.substr(begin, loc - begin);
+    if (format.at(loc + 1) == '%') {
+      ss << '%';
+      begin = loc + 2;
```
Magic numbers probably deserve a comment

gmagogsfm commented on torch/csrc/jit/runtime/vararg_functions.cpp:
```diff
@@ -39,6 +39,130 @@ void format(Stack& stack, size_t num_inputs) {
+      break;
+    }
+    default: {
+      TORCH_CHECK(
```
This error is not tested in your testcases.

gmagogsfm replied:
Also I'd recommend adding a percent sign before key so that it is clearly about formatting.

Commit:
Update on "Support %-based string formatting"




[ghstack-poisoned]

Commit:
Update on "Support %-based string formatting"




[ghstack-poisoned]

gmagogsfm commented on test/jit/test_string_formatting.py:
```diff
@@ -0,0 +1,156 @@
+        @torch.jit.script
+        def fn(arg1: str):
+            return "%s in template" % arg1
+        self.assertEqual("foo in template", fn("foo"))
```
Is TorchScripted percentFormat generating exactly same output as native python? It would be better if we can use `self.checkScript` in these tests to ensure eager and TS behavior match.

Commit:
Update on "Support %-based string formatting"




[ghstack-poisoned]

Commit:
Update on "Support %-based string formatting"




[ghstack-poisoned]

Commit:
Update on "Support %-based string formatting"




[ghstack-poisoned]

Commit:
Update on "Support %-based string formatting"




[ghstack-poisoned]

Commit:
Update on "Support %-based string formatting"




[ghstack-poisoned]

Commit:
Update on "Support %-based string formatting"




[ghstack-poisoned]

Commit:
Update on "Support %-based string formatting"


Differential Revision: [D24374215](https://our.internmc.facebook.com/intern/diff/D24374215)

[ghstack-poisoned]
